### DISCUSSION TITLE
Representing unknown dims as a symbol ?

### DISCUSSION THREAD

**Discussion by alexander-camuto** [sonos/tract#1259] [General]
Have a small model exported from sklearn (for which it is hard to create symbolic dims etc...) and the variable batch dim is represented as unknown in onnx (`?`) -- this makes it hard to infill in tract as it isn't explicitly represented as a symbol. Setting `?` to a value using the tract api doesn't work. 

Was wondering if it might make sense for the tract API to treat "?" as a symbol rather than as an unknown, such that models (such as the one zipped and attached) can compile. 

[network.onnx.zip](https://github.com/sonos/tract/files/13288657/network.onnx.zip)

**Comment by kali** [comment]
It actually sounds like a good idea. I'm gonna do some thinking :)

In the meatime, and for everyone who reads this, the workaround is to force symbol introdction by doing `-i SomeSymbol,3,f32` on the command line. On the rust API, it's a little bit more involved:

```
let mut model = tract_onnx().onnx().model_for_path("model.onnx")?;
let some_symbol = model.symbol_table.sym("SomeSymbol");
model.set_input_fact(0, f32::shape(dims!(some_symbol, 3))
[...]
```

The future "python/rust unified" API will also accept facts in string form, just as the command line does (it's crate tract-rs, but it's not ready for prime-time yet).

But yeah, I agree there may be value in having tract figuring it out on its own, at least for the trivial cases.
